crates/adapters/src/transport/nexmark.rs

```diff
@@ -0,0 +1,326 @@
+        Ok(())
+    }
+
+    fn disconnect(&self) {
+        self.inner.status[self.table].store(PipelineState::Terminated, Ordering::Release);
+        self.inner.unpark();
+    }
+}
+
+static INNER: Mutex<Weak<Inner>> = Mutex::new(Weak::new());
+
+struct Inner {
+    /// Status for each of the tables. We only run if all of them can run.
+    status: EnumMap<NexmarkTable, Atomic<PipelineState>>,
+
+    /// Options, which can be set from any of the tables but only from one of them.
+    options: OnceLock<NexmarkInputOptions>,
+
+    /// The per-table consumers.
+    consumers: Mutex<EnumMap<NexmarkTable, Option<Box<dyn InputConsumer>>>>,
+
+    /// The threads to wake up when we unpark.
+    ///
+    /// While we're waiting for the connectors for all the tables to be brought
+    /// to the running state for the first time, this is only the worker
+    /// thread. After that, we start the generator threads and they get
+    /// populated here instead.
+    threads: Mutex<Vec<Thread>>,
+}
+
+impl Inner {
+    pub fn new() -> Arc<Self> {
+        let inner = Arc::new(Self {
+            status: EnumMap::from_fn(|_| Atomic::new(PipelineState::Paused)),
+            options: OnceLock::new(),
+            consumers: Mutex::new(EnumMap::default()),
+            threads: Mutex::new(Vec::new()),
+        });
+        thread::Builder::new()
+            .name(String::from("nexmark"))
+            .spawn({
+                let inner = Arc::clone(&inner);
+                || inner.worker_thread()
+            })
+            .unwrap();
+        inner
+    }
+    pub fn merge(
+        &self,
+        config: &NexmarkInputConfig,
+        consumer: Box<dyn InputConsumer>,
+    ) -> AnyResult<()> {
+        let mut tables = self.consumers.lock().unwrap();
+        if tables[config.table].is_some() {
+            return Err(anyhow!(
+                "more than one Nexmark input connector for {:?}",
+                config.table
+            ));
+        }
+        tables[config.table] = Some(consumer);
+        drop(tables);
+
+        if let Some(options) = config.options.as_ref() {
+            if self.options.set(options.clone()).is_err() {
+                return Err(anyhow!(
+                    "can't configure Nexmark options from two different connectors"
+                ));
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn unpark(&self) {
+        for thread in self.threads.lock().unwrap().iter() {
+            thread.unpark();
+        }
+    }
+
+    /// Returns the pipeline's overall status based on the three underlying connectors:
+    ///
+    /// - We're terminated if any of the connectors are terminated.
+    /// - Otherwise, we're paused if any of them are paused.
+    /// - Otherwise, all of them are running, so we are running.
+    ///
+    /// To be honest, we could probably just use the status of the `bid`
+    /// connector, because it's 92% of the records, and ignore the rest.
+    fn status(&self) -> PipelineState {
```

Comment:
normally bids and auctions are interleaved, I don't know what the generator really does

Reply:
especially if we annotate the tables with foreign keys, it is important to have the foreign keys inserted after the corresponding keys have been

Reply:
I think that each thread generates an independent stream. Because of the dependency graph of the tables, that means that we should be OK if, from each batch, we first insert people, then auctions, then bids.